### PR TITLE
Add indentation indicator and selector

### PR DESCRIPTION
**Pull request by Abdillah**
The indicator include tile for customizing `tabType` and `tabLength`. The tile purposely using toggle link interface to switch between option.

Trying to solve issue #36.

### CODE REVIEW

**Comment by Abdillah** [comment]
__Notes__: The checks failed because this pull request based on `atom:master` branch, especially commit https://github.com/atom/atom/commit/0608f9438707e5383ddc540d0ba1aa64a8a3f28f.

But, I need some feedback on
1. The use of toggle interface (where you change `tabType` and `tabLength` just by click on the link) instead of SelectionListView (like command pallete).
2. Icon and design of it (currently I use `»` and lowercase letter) e.g. `» space 2`, `» tab 4`.

**Comment by mnquintana** [comment]
> Icon and design of it (currently I use » and lowercase letter) e.g. » space 2, » tab 4.

Can you post a screenshot of what you have currently?

**Comment by Abdillah** [comment]
Here some :camera: 

1. No buffer opened
screenshot from 2015-08-19 10-09-01

2. Full Window
screenshot from 2015-08-19 10-14-28

3. Buffer opened, autodetect
screenshot from 2015-08-19 10-09-19

4. Mouse over it. When click, it toggled
screenshot from 2015-08-19 10-09-40

I admit this far too simple and so I need feedback.

**Comment by simurai** [comment]
Spaces could have 2 dots as an "icon":

`‥ Space 2` 
`» Tab 4`

or between the text and number

`Space ‥ 2` 
`Tab » 4`

**Comment by Abdillah** [comment]
@simurai I like your suggestions, but... Atom draws `‥` as plain box (like `[]`). I suspect because it is longer than 8 bit.

How about octicons `icon-primitive-dot` and `icon-chevron-right`? While not quite right in semantic, it clear at the first sight.

**Comment by mnquintana** [comment]
> How about octicons icon-primitive-dot and icon-chevron-right? While not quite right in semantic, it clear at the first sight.

I like this :+1: - I think they'd also look more like icons than using the characters directly.

**Comment by simurai** [comment]
Using Octicons sounds good.

Or if it's unclear and space is a concern, no icons and just have the text labels might be already enough.

**Comment by Abdillah** [comment]
Here another shoot :camera:, this time with octicons.

### Full Window
screenshot from 2015-08-20 10-06-07
screenshot from 2015-08-20 10-06-13

### Status Barr
__Note:__
I add edge brackets to make it wider thus easing mouse focus and click on `tabLength` selector.

screenshot from 2015-08-20 10-05-02
screenshot from 2015-08-20 10-05-20
screenshot from 2015-08-20 10-05-35

For me, it looks no problem and visually good..

Ready for push!

**Comment by simurai** [comment]
Looking at this again, how about not having icons at all?

The chevron might make it feel like it connects with the word on the left: `CoffeScript   > Tab [4]`
And a dot is usually used to indicate a status.

So maybe the guideline for the status bar could be to only use icons if they are super clear. Like the merge icon. I think having the words `Tab` and `Space` should already be enough to understand its meaning.

**Comment by Abdillah** [comment]
Well then, I will remove those...

**Comment by Abdillah** [comment]
Silly, I push the commit to the `master` initially instead of this `abd-indent-selector` branch. 

Sorry for too long delay..

**Comment by Alhadis** [comment]
I'd pluralise the label, personally.

"Tabs [ 4 ]" makes it more obvious that the label isn't referring to the fourth tab the user currently has open.

Since the status bar is a graphical interpretation of the current states of the document and workspace, it's also more precise from a grammatical standpoint. E.g.:

> This document is using **LF** terminators
> This document is using **UTF-8** encoding
> This document is using **Tabs [ 4 ]** because spaces are selfish

I confess I'm getting pedantic, but yeah.

File: lib/indentation-status-view.coffee [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+  getActiveTextEditor: ->
+    atom.workspace.getActiveTextEditor()
+
+  tabTypeToggle: (ev) ->
+    editor = @getActiveTextEditor()
+    # Get current value
+    softTabs = editor.getSoftTabs(editor.getRootScopeDescriptor())
+    # Toggle softTabs
+    editor.setSoftTabs(not softTabs, editor.getRootScopeDescriptor())
+    # console.log 'toggle softTabs to ' + (not softTabs)
```
Comment (joshaber) [inline lib/indentation-status-view.coffee]: Could these be removed now?

File: lib/indentation-status-view.coffee [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+    editor.setTabLength(tabWidth)
+    @updateTabWidthText()
+
+  subscribeToPotentialEvents: ->
+    @userChangeSubscription?.dispose()
+    @indentationConfigSubscription?.dispose()
+    @indentationConfigSubscription = atom.config.observe 'editor.tabType', =>
+      @updateTabTypeText()
+    # FIXME: Noisy channel subscription, change to better one
```
Comment (joshaber) [inline lib/indentation-status-view.coffee]: Indeed, `onDidStopChanging` is probably better.

File: lib/indentation-status-view.coffee [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    else
+      @hide()
+
+  updateTabWidthText: ->
+    editor = @getActiveTextEditor()
+    if editor
+      tabwidth = editor.getTabLength()
+    else
+      @hide()
+    # console.log 'update to ' + tabwidth
```
Comment (joshaber) [inline lib/indentation-status-view.coffee]: Can we remove this?

File: lib/indentation-status-view.coffee [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    @tabLengthLink.textContent = '[ ' + tabwidth + ' ]'
+
+module.exports = document.registerElement('indentation-status', prototype: IndentationStatusView.prototype)
```
Comment (joshaber) [inline lib/indentation-status-view.coffee]: We're trying to move away from custom elements. See https://github.com/atom/atom/issues/5503#issuecomment-188446070.

**Comment by joshaber** [comment]
@Abdillah Thanks for opening this! I've left some comments above.

**Comment by damieng** [comment]
Any interest in making these changes or is this PR abandoned?

**Comment by Abdillah** [comment]
Sorry, well, I'm busy with other thing so not paying attention to the notification.
I will do it hopefully in this weekend. I'm happy if anyone can help, I've enabled "Allow edits from maintainers.".

**Comment by bronson** [comment]
Have you seen https://github.com/lee-dohm/indentation-indicator ?

**Comment by bronson** [comment]
Ha, I didn't see https://github.com/atom/status-bar/issues/36 ...  Yes you have.  Sorry for the noise.

**Comment by mehcode** [comment]
Just want to say that I really prefer how https://atom.io/packages/tab-control looks visual wise. 

screen shot 2017-04-27 at 10 48 08 pm

**Comment by damieng** [comment]
Closing due to inactivity.

**Pull request closed without merging**